src/robot/libraries/Telnet.py

```diff
@@ -491,8 +531,17 @@ def __init__(self, host=None, port=23, timeout=3.0, newline='CRLF',
                  encoding='UTF-8', encoding_errors='ignore',
                  default_log_level='INFO', window_size=None, environ_user=None,
                  terminal_emulation=False, terminal_type=None,
-                 telnetlib_log_level='TRACE'):
-        telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+                 telnetlib_log_level='TRACE', connection_timeout=None):
+        if connection_timeout is None:
+            telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+        else:
+            try:
+                telnetlib.Telnet.__init__(self, host,
+                                          int(port) if port else 23,
+                                          connection_timeout)
+            except socket.timeout:
+                raise AssertionError("Connection timeout after %s."
+                                     % secs_to_timestr(connection_timeout))
```

Comment:
Makes sense to let the socket exception go through. In my environment the test passes as trying to connect to that address causes a timeout because the address does not exists. I'm running in a Ubuntu 14.04 VM (must confirm the python version). I had issues getting the timeout when passing a hostname to the init function, as the DNS lookup does not respect the timeout and issue another exception when the name is not found.

Anyway, I'll try to remove this and try to use you approach for testing.